File: backend/internal/modules/finance/split.go
package finance

import "math"

// Compute derives the nine split values from a month's two incomes and four
// rates. It is a DIRECT PORT of the retiring `fin` service's
// backend/internal/months/split.go and the formula is LOCKED (PRD §V6-4 FR-F3,
// D82): the arithmetic below was confirmed with Karel on 2026-06-10 and carried
// over unchanged. Only the package name differs from the original.
//
// THE ORDER IS THE SPECIFICATION, not an implementation detail:
//
//  1. personal is rounded FIRST, so personal + toOperational == income holds
//     exactly per person (the transfer is a subtraction, never a second rounding
//     of income × (1−p)).
//  2. the two savings transfers round off the TOTAL income, not off what the
//     operational account received.
//  3. needs is a SUBTRACTION — it is where every rounding error is deliberately
//     parked so both the per-person and the per-account totals reconcile.
//
// Any "tidy-up" that rounds needs directly, or that computes toOperational as
// round(income × (1−p)), silently breaks an invariant and will not be caught by
// the worked example alone. See TestInvariants.
//
// The four invariants (asserted in split_test.go over every fixture and over
// randomised inputs):
//
//	PersonalKaja + ToOperationalKaja == incomeKaja
//	PersonalAndy + ToOperationalAndy == incomeAndy
//	FunSavings + NoFunSavings + Needs == OperationalReceived
//	PersonalKaja + PersonalAndy + OperationalReceived == TotalIncome
//
// KNOWN EDGE CASE, inherited and deliberately preserved: because Needs is the
// remainder it can come out NEGATIVE, by at most 2 CZK. The bound is provable —
// personal rounding moves OperationalReceived by at most 1 and each savings
// rounding by at most ½, so Needs ≥ T×o/100 − 2, and therefore Needs < 0
// requires T × o < 200. That happens at operational = 0 for any income (worst
// case −2) and, for operational ≥ 1, only below 200 Kč of total income.
// DO NOT clamp it: clamping breaks FunSavings + NoFunSavings + Needs ==
// OperationalReceived, the invariant the whole rounding scheme exists to
// protect. The frontend renders a negative Needs as "0 Kč" with a footnote; the
// data stays honest.
//
// Money is whole CZK end to end. The only float64 in the system lives between
// the multiplication and math.Round below — exactly where `fin` put it.
func Compute(incomeKaja, incomeAndy int, r Rates) Split {
	total := incomeKaja + incomeAndy
	p := float64(r.Personal) / 100
	f := float64(r.Fun) / 100
	n := float64(r.NoFun) / 100

	personalKaja := int(math.Round(float64(incomeKaja) * p))
	personalAndy := int(math.Round(float64(incomeAndy) * p))

	toOperationalKaja := incomeKaja - personalKaja
	toOperationalAndy := incomeAndy - personalAndy
	opReceived := toOperationalKaja + toOperationalAndy

	fun := int(math.Round(float64(total) * f))
	noFun := int(math.Round(float64(total) * n))
	needs := opReceived - fun - noFun

	return Split{
		TotalIncome:         total,
		PersonalKaja:        personalKaja,
		PersonalAndy:        personalAndy,
		ToOperationalKaja:   toOperationalKaja,
		ToOperationalAndy:   toOperationalAndy,
		OperationalReceived: opReceived,
		FunSavings:          fun,
		NoFunSavings:        noFun,
		Needs:               needs,
	}
}

// Sum returns the four rates added together. Valid input sums to exactly 100
// (PRD FR-F2); the service rejects anything else with a 422 and the table's own
// CHECK is the second net.
func (r Rates) Sum() int { return r.Personal + r.Operational + r.Fun + r.NoFun }

// Savings is the pooled total that leaves the operational account each month.
// Savings are joint and are never attributed per person.
func (s Split) Savings() int { return s.FunSavings + s.NoFunSavings }

File: frontend/src/modules/finance/split.ts
import type { FinanceRates, FinanceSplit } from './api/types'

// A MIRROR of the locked split formula, not a second source of truth.
//
// It mirrors backend/internal/modules/finance/split.go (PRD §V6-4 FR-F3, D82),
// and it exists for exactly ONE job: previewing a split in the add/edit form
// before anything is saved, so someone changing a rate sees the money move. That
// is what makes the form worth using.
//
// Everything rendered in the month list, the flow visualisation and the Nástěnka
// widget comes from the SERVER's `split` block. If this file and the Go one ever
// disagree, the Go one is right — and split.test.ts, which runs the same worked
// example the Go test does, is what catches the drift.
//
// The order is the specification: personal rounds FIRST (so personal +
// to_operational === income per person), savings round off the TOTAL, and needs
// is a SUBTRACTION that absorbs every rounding error. needs can come out
// negative by up to 2 Kč; it is deliberately NOT clamped here — the UI decides
// to display 0 Kč with a footnote, which is a rendering choice, not a data one.
//
// JS Math.round is half-UP (−0.5 → −0), Go's math.Round is half-AWAY-from-zero.
// The two agree for every input this formula can produce, because incomes are
// non-negative and the rates are non-negative percentages, so no product is ever
// negative.
export function computeSplit(incomeKaja: number, incomeAndy: number, rates: FinanceRates): FinanceSplit {
  const total = incomeKaja + incomeAndy
  const p = rates.personal / 100
  const f = rates.fun / 100
  const n = rates.no_fun / 100

  const personalKaja = Math.round(incomeKaja * p)
  const personalAndy = Math.round(incomeAndy * p)

  const toOperationalKaja = incomeKaja - personalKaja
  const toOperationalAndy = incomeAndy - personalAndy
  const operationalReceived = toOperationalKaja + toOperationalAndy

  const funSavings = Math.round(total * f)
  const noFunSavings = Math.round(total * n)
  const needs = operationalReceived - funSavings - noFunSavings

  return {
    total_income: total,
    personal_kaja: personalKaja,
    personal_andy: personalAndy,
    to_operational_kaja: toOperationalKaja,
    to_operational_andy: toOperationalAndy,
    operational_received: operationalReceived,
    fun_savings: funSavings,
    no_fun_savings: noFunSavings,
    needs,
  }
}

/** The four rates added together. The form blocks submit until this is 100. */
export function ratesSum(r: FinanceRates): number {
  return r.personal + r.operational + r.fun + r.no_fun
}

/** Pooled savings — fun + no-fun. Never attributed per person. */
export function savingsOf(s: FinanceSplit): number {
  return s.fun_savings + s.no_fun_savings
}
